v3/lints/cabf_ev/lint_ev_orgid_inconsistent_subj_and_ext.go

```diff
@@ -0,0 +1,114 @@
+/*
+ * ZLint Copyright 2024 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * Contributed by Adriano Santoni <adriano.santoni@staff.aruba.it>
+ * of ACTALIS S.p.A. (www.actalis.com).
+ */
+
+package cabf_ev
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+
+	"errors"
+	"regexp"
+)
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_ev_orgid_inconsistent_subj_and_ext",
+			Description:   "Checks that the organizationIdentifier Subject attribute and the CABFOrganizationIdentifier extension are consistent",
+			Citation:      "EVGs 9.2.8 and 9.8.2",
+			Source:        lint.CABFEVGuidelines,
+			EffectiveDate: util.CABFEV_Sec9_2_8_Date,
+		},
+		Lint: NewOrgIdInconsistentSubjAndExt,
+	})
+}
+
+// According to EVGs 9.2.8
+type OrganizationIdentifier struct {
+	Scheme    string
+	Country   string
+	State     string
+	Reference string
+}
+
+func ParseOrgId(orgIdString string, orgId *OrganizationIdentifier) error {
+
+	// This is according to the EVG (stricter than ETSI EN 319 412-1)
+	OrgIdPattern := `^[A-Z]{3}[A-Z]{2}(?:\+[A-Z]{2})?\-.+$`
+
+	compiledRegexp, err := regexp.Compile(OrgIdPattern)
+	if err != nil {
+		// This should neve occur, but one never knows....
+		panic(err)
+	}
+
+	if !compiledRegexp.MatchString(orgIdString) {
+		return errors.New("Cannot parse organizationIdentifier: it is probably invalid")
+	}
+
+	orgId.Scheme = orgIdString[0:3]
```

Comment:
I'm always torn on this because

1. UTF8 is trickier than it seems
2. It at least _seems_ like all of our data is ASCII, despite not being declared as such,

For the first point, consider the [following example](https://go.dev/play/p/Nvci4QaO2kQ).

```go
package main

import (
	"fmt"
)

func main() {
	var darnit string = "Hello, 世界"
	// You would think it would be 世 but we get ä
	fmt.Println(string(darnit[7]))
}
```

This is because indexing into a string does a literal lookup to the byte at that location, however UTF8 is a variable length encoding.

For the second point, the EV BRs allows for this to be a `UTF8String`. However the contents seem to be entirely sources from ISO 3166, and those [seem to all be ASCII](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) without explicitly declaring as such. Additionally, the test corpus passes and it is fairly representative of what is out in the wild.

---

If you care to consider this then casting to a `[]rune` before indexing does the right thing.

```go
string([]rune(orgIdString)[6:8])
```

Note that, although it is not an issue in _this_ scenario, but do make a mental note that the above is linear time and may result in an allocation. That would be a nasty surprise sometime in the future.

Reply:
I think using named capture groups would resolve the UTF-8 concern. It would also have the benefit of  removing the "magic string indexes", improving clarity of the code.